### PR TITLE
netbox: 3.3.9 -> 3.4.7, netbox_3_3: init at 3.3.10, RFC42-style options, more tests

### PR DESCRIPTION
**Pull request by minijackson**
###### Description of changes

- [3.4.0 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.0) (with breaking changes)
- [3.4.1 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.1)
- [3.4.2 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.2)
- [3.4.3 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.3)
- [3.4.4 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.4)
- [3.4.5 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.5)

Also added RFC42-style options, which introduces a new `pythonVars` format which might be interesting for other Django apps.

Also extended the test, since we missed the Django 4.1 incompatibility / LDAP integration issue earlier.

Added myself as maintainer, since I will be probably using it for quite some time ^^

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@RaitoBezarius @n0emis

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Thank you for this amazing change, I wanted to revisit our options since a long time, thank you for doing it. I hope to take a look to this in the next week.

**Comment by centromere** [comment]
3.4.3 has been [released](https://github.com/netbox-community/netbox/releases/tag/v3.4.3).

@RaitoBezarius, are you able to give this a review?

**Comment by minijackson** [comment]
Upgraded to 3.4.3, and rebased due to release note conflicts

File: pkgs/servers/web-apps/netbox/default.nix [hunk 1]
```diff
@@ -18,26 +18,20 @@ let
...
       owner = "netbox-community";
       repo = pname;
       rev = "refs/tags/v${version}";
-      sha256 = "sha256-KhnxD5pjlEIgISl4RMbhLCDwgUDfGFRi88ZcP1ndMhI=";
+      sha256 = "sha256-M/CbnlnRaWuwcJJofY5mPpVTl9zYrApsXqV9hilNyvw=";
     };
 
     patches = [
       # Allow setting the STATIC_ROOT from within the configuration and setting a custom redis URL
       ./config.patch
```
Comment (RaitoBezarius) [inline pkgs/servers/web-apps/netbox/default.nix:34]: Did we try to upstream this?

**Review by RaitoBezarius — APPROVED** [review]
Well, all of this is very high quality. :)
My only worries are the following:

- should not we introduce a state version mechanism?
- should we introduce a flag to disable automatic migrations? (to prevent any dataloss)

File: pkgs/servers/web-apps/netbox/default.nix [hunk 2]
```diff
@@ -18,26 +18,20 @@ let
...
-      ./graphql-3_2_0.patch
-      # fix compatibility ith django 4.1
-      (fetchpatch {
-        url = "https://github.com/netbox-community/netbox/pull/10341/commits/ce6bf9e5c1bc08edc80f6ea1e55cf1318ae6e14b.patch";
-        sha256 = "sha256-aCPQp6k7Zwga29euASAd+f13hIcZnIUu3RPAzNPqgxc=";
-      })
     ];
 
     propagatedBuildInputs = with py.pkgs; [
```
Comment (RaitoBezarius) [inline pkgs/servers/web-apps/netbox/default.nix:37]: Let's backpropagate the Netbox tests here?

**Comment by RaitoBezarius** [comment]
@minijackson Can I do something to help you on this PR? (also, can we bump it to the latest version?)

File: pkgs/servers/web-apps/netbox/default.nix (hunk 1 above)
Comment (minijackson) [reply pkgs/servers/web-apps/netbox/default.nix:34]: Not yet, but I think we should

File: pkgs/servers/web-apps/netbox/default.nix (hunk 2 above)
Comment (minijackson) [reply pkgs/servers/web-apps/netbox/default.nix:37]: You mean `passthru.tests.netbox`? It should be already there, the tests only were expanded.

**Comment by minijackson** [comment]
> @minijackson Can I do something to help you on this PR? (also, can we bump it to the latest version?)

On it! I just had to have some time to do it.

File: pkgs/servers/web-apps/netbox/default.nix (hunk 2 above)
Comment (RaitoBezarius) [reply pkgs/servers/web-apps/netbox/default.nix:37]: Got it

**Review by RaitoBezarius — APPROVED** [review]
The only problem I "still have" regarding this PR is the automatic migrations.
But it is hard to take a decision.
Well, we have explicit release notes about it, so this is only a weak blocker for me.
We should definitely ask what is upstream opinion on this now and if we need to introduce proper stateVersion and graceful upgrades.

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 206983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>netbox</li>
  </ul>
</details>

**Comment by RaitoBezarius** [comment]
Redis is failing to build for me:
```
> !!! WARNING The following tests failed:
       >
       > *** [err]: AOF will open a temporary INCR AOF to accumulate data until the first AOFRW success when AOF is dynamically enabled in tests/integration/aof-multi-part.tcl
       > Expected 'file appendonly.aof.2.base.rdb seq 2 type b' to be equal to 'file appendonly.aof.1.base.rdb seq 1 type b' (context: type source line 125 file /build/redis-7.0.9/tests/support/aofmanifest.tcl cmd {assert_equal [lindex $lines $i] [lindex $content $i]} proc ::assert_aof_manifest_content level 2)
       > Cleanup: may take some time... OK
       For full logs, run 'nix log /nix/store/68a2mzcsidqb8h4kz153c31c0drwj7hm-redis-7.0.9.drv'.
error: builder for '/nix/store/68a2mzcsidqb8h4kz153c31c0drwj7hm-redis-7.0.9.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/bkzshfakd85403dj97dh9c7ark208zm7-python3.10-django-redis-5.2.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/190g51x83kzjy362yaz7k8z8vhzj513j-netbox-3.4.5.drv' failed to build
error: 1 dependencies of derivation '/nix/store/fap8zc20jsq341yccf2f8ncywdd99gfm-review-shell.drv' failed to build
```

**Comment by minijackson** [comment]
That's weird, I just rebased, and ran `nix build '.#nixosTests.netbox'`. I had to rebuild the kernel and redis, but it built without issues...

**Comment by minijackson** [comment]
I can even `nix log /nix/store/68a2mzcsidqb8h4kz153c31c0drwj7hm-redis-7.0.9.drv` and it shows success

**Comment by RaitoBezarius** [comment]
> That's weird, I just rebased, and ran `nix build '.#nixosTests.netbox'`. I had to rebuild the kernel and redis, but it built without issues...

CI is working, so this might be a "me" thing, I was working on extremely cursed stuff just before, let me try to find a pure environment.

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 206983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netbox</li>
  </ul>
</details>

**Comment by RaitoBezarius** [comment]
Good, I think there's some flakiness in Redis tests but uncertain.
Anyway, what's your opinion on the migrations problem @minijackson ?

**Comment by minijackson** [comment]
It's kinda hard to have automatic migration testing, with the way things are built... I know quite a lot of services in NixOS that does automatic migration, without much issue, so my policy has always been "do automatic migration, gatekeep behing `stateVersion` on major upgrades, or when release notes mention something important"

**Comment by minijackson** [comment]
I have tried having a more conservative approach once, by leaving older versions, but it meant having unmaintained versions with potential security vulnerabilities available in nixpkgs :confused:

**Comment by RaitoBezarius** [comment]
> I have tried having a more conservative approach once, by leaving older versions, but it meant having unmaintained versions with potential security vulnerabilities available in nixpkgs confused

We can use `knownVulnerabilities` and let people know it's time to upgrade or they can choose to permit vulnerable versions. We can also backport when possible.

> It's kinda hard to have automatic migration testing, with the way things are built... I know quite a lot of services in NixOS that does automatic migration, without much issue, so my policy has always been "do automatic migration, gatekeep behing `stateVersion` on major upgrades, or when release notes mention something important"

Yes, I don't think automatic migration testing is feasible neither.

My question is regarding this specific package, do you think we should keep older version around for at least one release cycle and give the opportunity to users to upgrade gracefully?

**Comment by RaitoBezarius** [comment]
@minijackson I feel like reintroducing `netbox_3_2` and locking behind a `stateVersion` things is good for NixOS 22.11.
We can do it in a follow-up PR, does that work for you?

**Comment by minijackson** [comment]
welp, I thought it was a minor upgrade... But 3.4.0 had a breaking change. We should keep the old version around and use `stateVersion`, but I think we're in a special case, since the current packaged version doesn't really work... (see: https://github.com/netbox-community/netbox/discussions/11139)

I'm not a big fan of defaulting to our 3.3.9 version, and since the issue went unreported for a while, I think it's fine for this instance.

> @minijackson I feel like reintroducing `netbox_3_2` and locking behind a `stateVersion` things is good for NixOS 22.11. We can do it in a follow-up PR, does that work for you?

Isn't 3.3.9 already the version in 22.11? I'd be *very* wary of backwards migrations

**Comment by minijackson** [comment]
we could fix and reintroduce our 3.3.9 version, though, by also introducing Django 4.0.8

**Comment by RaitoBezarius** [comment]
My bad 3.3.9 is good yes

Le dim. 12 mars 2023 à 18:18, Rémi NICOLE ***@***.***> a
écrit :

**Review by n0emis — APPROVED** [review]
Thanks for your work! Nextcloud 3.4.6 was released yesterday, would be be able to update your PR? otherwise we can do this seperately to speed things up a bit...

**Comment by minijackson** [comment]
@n0emis done! Tested with the netbox tests locally, and they pass

**Review by n0emis — APPROVED** [review]

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Things remaining to do:

- stateVersion locking; previous netbox version,
- try to upstream STATIC_ROOT patch

**Comment by minijackson** [comment]
>     * stateVersion locking; previous netbox version,

Done, still needs to be fixed in another PR

>     * try to upstream STATIC_ROOT patch

No matter what we do, this will end up in a newer version, I think it's best to do this separately from this PR

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 3]
```diff
@@ -172,6 +172,10 @@ In addition to numerous new and upgraded packages, this release has the followin
   - `services.openssh.ciphers` to `services.openssh.settings.Ciphers`
   - `services.openssh.gatewayPorts` to `services.openssh.settings.GatewayPorts`
 
+- `netbox` was upgraded from the 3.3 branch to the 3.4 branch. Please make sure to check the breaking changes mentioned in the [3.4.0 release notes](https://github.com/netbox-community/netbox/releases/tag/v3.4.0)
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2305.section.md:175]: this should be removed?

**Comment by RaitoBezarius** [comment]
> > ```
> > * stateVersion locking; previous netbox version,
> > ```
> 
> Done, still needs to be fixed in another PR

why needs to be fixed in another PR?

> 
> > ```
> > * try to upstream STATIC_ROOT patch
> > ```
> 
> No matter what we do, this will end up in a newer version, I think it's best to do this separately from this PR

I meant putting a link to some issue or PR.

File: nixos/doc/manual/release-notes/rl-2305.section.md (hunk 3 above)
Comment (minijackson) [reply nixos/doc/manual/release-notes/rl-2305.section.md:175]: Yes, I completely forgot about it

**Comment by RaitoBezarius** [comment]
Regarding the last commit message:
> `netbox_3_3: reintroduce, use in NixOS if stateVersion < 23.05`
please put `netbox_3_3: init` and put the reintroduction message in the body and the explanation about stateVersion there

**Comment by minijackson** [comment]
It seems that I cannot reproduce the error that I got with NetBox 3.3.9, although the DJango dependency mismatch is still here. I tried running the NetBox test, and testing it manually in a container, and it works without issues.

**Comment by KiruyaMomochi** [comment]
When running the latest `nixpkgs` with patch of this PR, I can't start netbox successfully.

```
Mar 22 08:24:33 carmina systemd[1]: Starting NetBox migrations...
Mar 22 08:24:35 carmina netbox[74931]: Operations to perform:
Mar 22 08:24:35 carmina netbox[74931]:   Apply all migrations: admin, auth, circuits, contenttypes, dcim, django_rq, extras, ipam, sessions, social_django, taggit, tenancy, users, virtualization, wireless
Mar 22 08:24:36 carmina netbox[74931]: Running migrations:
Mar 22 08:24:36 carmina netbox[74931]:   No migrations to apply.
Mar 22 08:24:36 carmina netbox[74931]:   Your models in app(s): 'social_django' have changes that are not yet reflected in a migration, and so won't be applied.
Mar 22 08:24:36 carmina netbox[74931]:   Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.
Mar 22 08:24:36 carmina systemd[1]: netbox-migration.service: Deactivated successfully.
Mar 22 08:24:36 carmina systemd[1]: Finished NetBox migrations.
Mar 22 08:24:36 carmina systemd[1]: netbox-migration.service: Consumed 2.723s CPU time, no IP traffic.
```

I guess this is because [social-auth-app-django](https://github.com/NixOS/nixpkgs/pull/221647) is just got updated. We may need to pin an older version in the package file.

I have tried following [NixOS wiki](https://nixos.wiki/wiki/Overlays) to override `social-auth-app-django`. However, neither overriding `netbox` or just override `python3` package works for me.

**Comment by stuebinm** [comment]
> I have tried following [NixOS wiki](https://nixos.wiki/wiki/Overlays) to override `social-auth-app-django` in an overlay. However, neither overriding `netbox` or just override `python3` package works for me.
> 
> Is this because `python3` has already be overriden by this package?
>
> https://github.com/NixOS/nixpkgs/blob/297c4f66bafbe65e8b27fe914c9aaa625bce2a18/pkgs/servers/web-apps/netbox/default.nix#L11-L15

It is, this is a limitation of how python's `packageOverrides` work in `nixpkgs`. I've run into this issue before, and just opened https://github.com/NixOS/nixpkgs/pull/223268 with a workaround to allow overrides to work as expected.

Until then, an easy but hacky workaround (without changing nixpkgs) is to also put `django = super.django_4` into your own package overrides, and then disable python's overriding functionality entirely so the snippet you linked can't change it back (e.g. by doing `let python3 = (python3.override { (your package overrides) }) // { override = _: python3 } in ...`).

**Comment by RaitoBezarius** [comment]
3.4.7 is out, can I do something to help you get to the finish line here @minijackson ? I can take over and push the last changes if needed.

File: pkgs/servers/web-apps/netbox/3.3.nix [hunk 4]
```diff
@@ -0,0 +1,114 @@
+{ lib
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/netbox/3.3.nix:1]: Please do t duplicate the entire file

**Comment by RaitoBezarius** [comment]
@minijackson friendly ping so that I can take over if you're too busy right now :)

**Comment by minijackson** [comment]
@RaitoBezarius sure, go ahead. Sorry for the delay...

**Comment by RaitoBezarius** [comment]
@minijackson @SuperSandro2000 please could you give a review to this?

**Comment by minijackson** [comment]
Seems good, thanks! (Can't make this a GitHub approval, since I'm the original author of the PR)

All tests pass on my machine:

```bash
nix build -f ./nixpkgs "netbox.tests.netbox"
NIXPKGS_ALLOW_INSECURE=1 nix build -f ./nixpkgs "netbox_3_3.tests.netbox"
```

---

Result of `nixpkgs-review pr 206983` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>netbox_3_3</li>
  </ul>
</details>
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netbox</li>
  </ul>
</details>

**Comment by RaitoBezarius** [comment]
For the next time, let's try to do smaller PRs and smaller scale changes to get easier merges. :)
Thank you to everyone involved!

**Pull request merged**